### PR TITLE
refactor(api): move data plane select functionality to control api

### LINKED ISSUE #4177: Move data plane selection endpoints from management to control api

### PR DESCRIPTION
**Pull request by ndr-brt**
## What this PR changes/adds

Add `select` and `getAll` functionality to dataplane selector `control-api`, and deprecate `select` on the `management-api`

## Why it does that

Separation of responsibilities.

## Further notes

- made all methods of `DataPlaneSelectorService` return `ServiceResult`, this makes the interaction of the `Remote` implementation more consistent, because there could be situations (unlike with the `Embedded` one) where the communication with the server fails.
- removed `destination` from the dataplane selection request, because it is not really necessary (source and transferType are enough)
- in the `DataPlaneSignalingFlowController` I left a TODO for a refactoring about the way the DataPlaneClient is created (in my opinion dataplaneselection should be encapsulated in the factory, so the condition "is data plane embedded" will be more explicit and testable

## Linked Issue(s)

Closes #4177

_Please be sure to take a look at the [contributing guidelines](https://github.com/eclipse-edc/.github/blob/main/CONTRIBUTING.md#submit-a-pull-request) and our [etiquette for pull requests](https://github.com/eclipse-edc/.github/blob/main/contributing/pr_etiquette.md)._

### CODE REVIEW

File: extensions/control-plane/transfer/transfer-data-plane-signaling/src/main/java/org/eclipse/edc/connector/controlplane/transfer/dataplane/flow/DataPlaneSignalingFlowController.java [hunk 1]
```diff
@@ -97,46 +99,66 @@ public boolean canHandle(TransferProcess transferProcess) {
...
+            return clientFactory.createClient(dataPlaneInstance)
+                    .start(dataFlowRequest)
+                    .map(it -> DataFlowResponse.Builder.newInstance()
+                            .dataAddress(it.getDataAddress())
+                            .dataPlaneId(dataPlaneInstance.getId())
+                            .build()
+                    );
+        } else {
+            // TODO: this branch works for embedded data plane but it is a potential false positive when the dataplane is not found, needs to be refactored
+            return clientFactory.createClient(null)
```
Comment (paullatzelsperger) [inline extensions/control-plane/transfer/transfer-data-plane-signaling/src/main/java/org/eclipse/edc/connector/controlplane/transfer/dataplane/flow/DataPlaneSignalingFlowController.java:114]: shouldn't this just return a `failure`? It doesn't really matter whether no data plane exists to serve the request, or no data plane is _reachable_ that could serve the request.

File: extensions/control-plane/transfer/transfer-data-plane-signaling/src/main/java/org/eclipse/edc/connector/controlplane/transfer/dataplane/flow/DataPlaneSignalingFlowController.java [hunk 2]
```diff
@@ -97,46 +99,66 @@ public boolean canHandle(TransferProcess transferProcess) {
...
+        }
+
+        return result.getContent().stream()
                 .filter(it -> it.getAllowedSourceTypes().contains(asset.getDataAddress().getType()))
                 .map(DataPlaneInstance::getAllowedTransferTypes)
                 .flatMap(Collection::stream)
                 .collect(toSet());
     }
 
+    private StatusResult<Void> onDataplaneInstancesDo(String name, TransferProcess transferProcess, BiFunction<DataPlaneClient, String, StatusResult<Void>> action) {
```
Comment (paullatzelsperger) [inline extensions/control-plane/transfer/transfer-data-plane-signaling/src/main/java/org/eclipse/edc/connector/controlplane/transfer/dataplane/flow/DataPlaneSignalingFlowController.java:148]: micro-nit: would rename `name -> action`

File: extensions/data-plane-selector/data-plane-selector-api/src/main/java/org/eclipse/edc/connector/dataplane/selector/api/v2/DataplaneSelectorApiController.java [hunk 3]
```diff
@@ -98,7 +98,11 @@ public JsonObject addDataPlaneInstance(JsonObject jsonObject) {
     @Override
     @GET
     public JsonArray getAllDataPlaneInstances() {
-        var instances = selectionService.getAll();
+        var result = selectionService.getAll();
+        if (result.failed()) {
```
Comment (paullatzelsperger) [inline extensions/data-plane-selector/data-plane-selector-api/src/main/java/org/eclipse/edc/connector/dataplane/selector/api/v2/DataplaneSelectorApiController.java:102]: nit: `orElseThrow`?

File: extensions/data-plane-selector/data-plane-selector-control-api/src/test/java/org/eclipse/edc/connector/dataplane/selector/control/api/DataplaneSelectorControlApiControllerTest.java [hunk 4]
```diff
@@ -139,6 +143,156 @@ void shouldFail_whenEgressTransformationFails() {
...
+                    .build();
+            when(service.getAll()).thenReturn(ServiceResult.success(List.of(dataPlane)));
+            when(typeTransformerRegistry.transform(any(), eq(JsonObject.class)))
+                    .thenReturn(Result.failure("error"));
+
+            given()
+                    .port(port)
+                    .get("/v1/dataplanes")
+                    .then()
+                    .statusCode(200)
```
Comment (paullatzelsperger) [inline extensions/data-plane-selector/data-plane-selector-control-api/src/test/java/org/eclipse/edc/connector/dataplane/selector/control/api/DataplaneSelectorControlApiControllerTest.java:279]: in other APIs we have this as a 500, for example an egress transformation error on selection (Line 218)

File: extensions/control-plane/transfer/transfer-data-plane-signaling/src/main/java/org/eclipse/edc/connector/controlplane/transfer/dataplane/flow/DataPlaneSignalingFlowController.java (hunk 1 above)
Comment (ndr-brt) [reply extensions/control-plane/transfer/transfer-data-plane-signaling/src/main/java/org/eclipse/edc/connector/controlplane/transfer/dataplane/flow/DataPlaneSignalingFlowController.java:114]: needs to work when the data plane is embedded, that's why I left a comment, it is covered by e2e tests, when the dataplane is embedded the selector returns null (so not-found), already on my todo list.

File: extensions/control-plane/transfer/transfer-data-plane-signaling/src/main/java/org/eclipse/edc/connector/controlplane/transfer/dataplane/flow/DataPlaneSignalingFlowController.java (hunk 2 above)
Comment (ndr-brt) [reply extensions/control-plane/transfer/transfer-data-plane-signaling/src/main/java/org/eclipse/edc/connector/controlplane/transfer/dataplane/flow/DataPlaneSignalingFlowController.java:148]: ok, I'll need to find a different name for `action` then :joy:

File: extensions/data-plane-selector/data-plane-selector-control-api/src/test/java/org/eclipse/edc/connector/dataplane/selector/control/api/DataplaneSelectorControlApiControllerTest.java (hunk 4 above)
Comment (paullatzelsperger) [reply extensions/data-plane-selector/data-plane-selector-control-api/src/test/java/org/eclipse/edc/connector/dataplane/selector/control/api/DataplaneSelectorControlApiControllerTest.java:279]: yep, that's right. I kinda lean towards always throwing an exception, but tackling this globally in a subsequent cleanup issue/pr makes sense.

File: extensions/control-plane/transfer/transfer-data-plane-signaling/src/main/java/org/eclipse/edc/connector/controlplane/transfer/dataplane/flow/DataPlaneSignalingFlowController.java (hunk 2 above)
Comment (paullatzelsperger) [reply extensions/control-plane/transfer/transfer-data-plane-signaling/src/main/java/org/eclipse/edc/connector/controlplane/transfer/dataplane/flow/DataPlaneSignalingFlowController.java]: ah shit, overlooked that :D

File: extensions/control-plane/transfer/transfer-data-plane-signaling/src/main/java/org/eclipse/edc/connector/controlplane/transfer/dataplane/flow/DataPlaneSignalingFlowController.java (hunk 1 above)
Comment (paullatzelsperger) [reply extensions/control-plane/transfer/transfer-data-plane-signaling/src/main/java/org/eclipse/edc/connector/controlplane/transfer/dataplane/flow/DataPlaneSignalingFlowController.java:114]: ok good then.

**Review by paullatzelsperger — APPROVED** [review]

**Pull request merged**